src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerObserver.cs

```diff
@@ -38,9 +40,31 @@ public Task OpenAsync(ChangeFeedObserverContext context)
             return Task.CompletedTask;
         }
 
-        public Task ProcessChangesAsync(ChangeFeedObserverContext context, IReadOnlyList<Document> docs)
+        public async Task ProcessChangesAsync(ChangeFeedObserverContext context, IReadOnlyList<Document> docs)
         {
-            return this.executor.TryExecuteAsync(new TriggeredFunctionData() { TriggerValue = docs }, CancellationToken.None);
+            int retries = 0;
+
+            while (true)
+            {
+                var result = await this.executor.TryExecuteAsync(new TriggeredFunctionData() { TriggerValue = docs }, CancellationToken.None);
+
+                if (result.Succeeded)
+                {
+                    // notification processed successfully
+                    return;
+                }
+
+                // note: -1 means "infinite retries", so ignoring the count of retries in that case
+                if (retryCount != -1 && retries >= retryCount)
```

Comment:
What was the scenario for allowing infinite retries? Anything infinite always scares us, from a supportability standpoint  :-)

Reply:
There are I believe many cases where a system must process in-order and reliably.  At Least Once processing.  This can be critical for many use cases.  Since Documents may be updated incrementally, it may compromise downstream data integrity to "skip" something.  Even with a Dead Letter Queue (which is a good idea and an excellent add-on to this change), reverse-engineering the change sequence to "fix up" downstream dependencies, trying to "insert the missed update" may be difficult, requiring a lot of extra design and implementation effort, or even be effectively impossible.  

For our use case and I believe many others, it is better to see an error logged, repeatedly, until someone fixes it and unblocks the system.  Otherwise our data processing becomes corrupt.  One excellent example which you may be indirectly familiar with is using the change feed in order to maintain a cache of "what device ids are in our document store" and not processing NGP delete requests beyond the watermark of that cache.  Doing a costly self-join query on every NGP delete request would be prohibitive given the size of our collection.

There are several email threads (plus inline in this pull request) discussing this problem with your Architect, and further enumerating the reasons and justification.

Reply:
> Thanks @nkinnan. Sorry that this got punted for so long. Now that v2 has hit GA, I'll see what I can do to get this pushed over the finish line. I can handle the test-writing and the last few cleanup issues.

Much appreciated!  I'm happy to contribute however I can.

> I've asked @fabiocav and @mathewc to take a look at the approach as well. I want to make sure that whatever we do here looks good in a broader sense. For example, we may want to do the same thing for Event Hubs (since the behavior is similar) so we need to make sure we're not creating conflicting designs for the different triggers.

Absolutely!  All necessary care should be taken to ensure the design is as forward looking and consistent as possible.  You brought up the possibility of a Dead Letter Queue for the change feed which I think would be an excellent addition.  Just bringing that up as part of the consideration if you have existing DLQ functionality for EventHub and how that would fit in with a retry knob on the EH side.

Reply:
Also please remember in the "At Least Once" context, that the reason for failure may be downstream dependencies.  For example, I am writing information related to the change to an Azure Table, and Storage goes down for a few minutes.  Changes are distributed to the change feed processor At Least Once already, but that loses meaning when the processor function can't actually process the change, and that may be out of the function's control.

Reply:
@ealsur raised a concern (internally) that long retry loops could cause the lease to expire as the ChangeFeedProcessor would never renew the lease on the partition. That means another instance could start processing these same documents, leading to duplicate processing.

We're contacting the ChangeFeedProcessor owners to see if they have a better recommendation or if we can get better control in a future release.

I see the valid scenario -- but we want to make sure that all the pieces we're using can handle an infinite retry like this. Will report back soon.